From fa042f71096c5fb43befb47228a61b769a4dee40 Mon Sep 17 00:00:00 2001
From: "Yi, Hongyu" <hongyu.yi@intel.com>
Date: Thu, 14 May 2015 10:35:35 +0800
Subject: [PATCH 394/462] atomisp2: Enable ftrace IPU C/P state and utilization

Add new tracepoints "ipu_cstate" and "ipu_pstate" to track IPU
power state, frequency and utilization.

The output looks like:
Binder_1-2974 [002] ...1 476.638118: ipu_cstate:cstate=1
Binder_1-2974 [002] ...1 476.638369:ipu_pstate:freq=100 util=-1

"ipu_cstate: state=0" means IPU is suspended,
"ipu_cstate: state=1" means IPU is resumedi.
"ipu_pstate: freq=100 util=-1" means IPU freq is 100MHz, util
is invalid.

These ftraces can also be visulized by using tool like ux-tune.

Change-Id: Id123049aea29dab696c3d5c520a40cae98b431da
Tracked-on: https://jira01.devtools.intel.com/browse/IMINAN-36335
Signed-off-by: Yi, Hongyu <hongyu.yi@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/55934
Build: IMIN (SC)
Validation-Android: Wei Tao <wei.tao@intel.com>
Integrator: WeiX Meng <weix.meng@intel.com>
Build: WeiX Meng <weix.meng@intel.com>
Reviewed-by: WeiX Meng <weix.meng@intel.com>
Maintainer: WeiX Meng <weix.meng@intel.com>
Approver: WeiX Meng <weix.meng@intel.com>
---
 .../pci/atomisp2/atomisp_driver/atomisp_cmd.c      |  7 +++-
 .../atomisp2/atomisp_driver/atomisp_trace_event.h  | 42 +++++++++++++++++++++-
 .../pci/atomisp2/atomisp_driver/atomisp_v4l2.c     |  9 +++--
 3 files changed, 54 insertions(+), 4 deletions(-)

diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c
index be07e98..3de7a44 100755
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c
@@ -39,6 +39,9 @@
 #include <media/v4l2-event.h>
 #include <media/videobuf-vmalloc.h>
 
+#define CREATE_TRACE_POINTS
+#include "atomisp_trace_event.h"
+
 #include "atomisp_cmd.h"
 #include "atomisp_common.h"
 #include "atomisp_fops.h"
@@ -339,8 +342,10 @@ done:
 	dev_dbg(isp->dev, "Programming DFS frequency to %d\n", new_freq);
 
 	ret = write_target_freq_to_hw(isp, new_freq);
-	if (!ret)
+	if (!ret) {
 		isp->sw_contex.running_freq = new_freq;
+		trace_ipu_pstate(new_freq, -1);
+	}
 	return ret;
 }
 
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_trace_event.h b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_trace_event.h
index 60c1845..5ce282d 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_trace_event.h
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_trace_event.h
@@ -21,6 +21,9 @@
 #undef TRACE_SYSTEM
 #define TRACE_SYSTEM atomisp
 
+#if !defined(ATOMISP_TRACE_EVENT_H) || defined(TRACE_HEADER_MULTI_READ)
+#define ATOMISP_TRACE_EVENT_H
+
 #include <linux/tracepoint.h>
 #include <linux/string.h>
 TRACE_EVENT(camera_meminfo,
@@ -85,9 +88,46 @@ TRACE_EVENT(camera_debug,
 		__entry->info)
 );
 
+TRACE_EVENT(ipu_cstate,
+
+		TP_PROTO(int cstate),
+
+		TP_ARGS(cstate),
+
+		TP_STRUCT__entry(
+			__field(int, cstate)
+		),
+
+		TP_fast_assign(
+			__entry->cstate = cstate;
+		),
+
+		TP_printk("cstate=%d", __entry->cstate)
+);
+
+TRACE_EVENT(ipu_pstate,
+
+		TP_PROTO(int freq, int util),
+
+		TP_ARGS(freq, util),
+
+		TP_STRUCT__entry(
+			__field(int, freq)
+			__field(int, util)
+		),
+
+		TP_fast_assign(
+			__entry->freq = freq;
+			__entry->util = util;
+		),
+
+		TP_printk("freq=%d util=%d", __entry->freq, __entry->util)
+);
+#endif
+
 #undef TRACE_INCLUDE_PATH
 #undef TRACE_INCLUDE_FILE
-#define TRACE_INCLUDE_PATH (.)
+#define TRACE_INCLUDE_PATH .
 #define TRACE_INCLUDE_FILE   atomisp_trace_event
 /* This part must be outside protection */
 #include <trace/define_trace.h>
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_v4l2.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_v4l2.c
index 0c74f03..8f4483a 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_v4l2.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_v4l2.c
@@ -47,6 +47,7 @@
 #include "atomisp_dfs_tables.h"
 #include "atomisp_drvfs.h"
 #include "hmm/hmm.h"
+#include "atomisp_trace_event.h"
 
 #include "hrt/hive_isp_css_mm_hrt.h"
 
@@ -453,8 +454,10 @@ int atomisp_mrfld_power_down(struct atomisp_device *isp)
 				reg_value);
 		/* wait until ISPSSPM0 bit[25:24] shows 0x3 */
 		if ((reg_value >> MRFLD_ISPSSPM0_ISPSSS_OFFSET) ==
-			MRFLD_ISPSSPM0_IUNIT_POWER_OFF)
+			MRFLD_ISPSSPM0_IUNIT_POWER_OFF) {
+			trace_ipu_cstate(0);
 			return 0;
+		}
 
 		if (time_after(jiffies, timeout)) {
 			dev_err(isp->dev, "power-off iunit timeout.\n");
@@ -497,8 +500,10 @@ int atomisp_mrfld_power_up(struct atomisp_device *isp)
 				reg_value);
 		/* wait until ISPSSPM0 bit[25:24] shows 0x0 */
 		if ((reg_value >> MRFLD_ISPSSPM0_ISPSSS_OFFSET) ==
-			MRFLD_ISPSSPM0_IUNIT_POWER_ON)
+			MRFLD_ISPSSPM0_IUNIT_POWER_ON) {
+			trace_ipu_cstate(1);
 			return 0;
+		}
 
 		if (time_after(jiffies, timeout)) {
 			dev_err(isp->dev, "power-on iunit timeout.\n");
-- 
1.9.1

